controllers/controller/workspacerouting/solvers/common.go

```diff
@@ -68,50 +75,69 @@ func getDiscoverableServicesForEndpoints(endpoints map[string]controllerv1alpha1
 	return services
 }
 
-func getServicesForEndpoints(endpoints map[string]controllerv1alpha1.EndpointList, meta WorkspaceMetadata) []corev1.Service {
-	if len(endpoints) == 0 {
-		return nil
+// GetServiceForEndpoints returns a single service that exposes all endpoints of given exposure types, possibly also including the discoverable types.
+func GetServiceForEndpoints(endpoints map[string]controllerv1alpha1.EndpointList, meta WorkspaceMetadata, includeDiscoverable bool, exposureType ...devworkspace.EndpointExposure) corev1.Service {
+	service := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      common.ServiceName(meta.WorkspaceId),
+			Namespace: meta.Namespace,
+			Labels: map[string]string{
+				config.WorkspaceIDLabel: meta.WorkspaceId,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: meta.PodSelector,
+			Type:     corev1.ServiceTypeClusterIP,
+		},
 	}
-	var servicePorts []corev1.ServicePort
-	for _, machineEndpoints := range endpoints {
-		for _, endpoint := range machineEndpoints {
-			isExposed := func(port int) bool {
-				for _, port := range servicePorts {
-					if port.TargetPort.IntVal == int32(endpoint.TargetPort) {
-						//port is already exposed
-						return true
-					}
-				}
-				return false
+
+	// "set" of ports that are still left for exposure
+	ports := map[int]struct{}{}
+	for _, es := range endpoints {
+		for _, endpoint := range es {
+			ports[endpoint.TargetPort] = struct{}{}
+		}
+	}
+
+	// "set" of exposure types that are allowed
+	validExposures := map[v1alpha2.EndpointExposure]struct{}{}
```

Comment:
Canonically, a set can be defined as `map[v1alpha2.EndpointExposure]bool` -- then you can use it as expected:
```golang
// Add to set:
validExposures[myExposure] = true
// Check membership:
if validExposures[myExposure] {
  // set contains myExposure
}
```
This uses the helpful fact that trying to read a non-existent key from a map returns the types zero value (which is `false` for bool)

Reply:
Interesting. The code is updated now.